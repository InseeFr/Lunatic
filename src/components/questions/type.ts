import { ReactNode } from 'react-markdown/lib/ast-to-react';

export type MessageType = {
	text?: ReactNode;
};
