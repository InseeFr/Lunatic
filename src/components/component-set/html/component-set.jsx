import React from 'react';
import { createCustomizableLunaticField, Errors } from '../../commons';
import Legend from './legend';

function ComponentSet({ id, errors, description, legendText, children }) {
	const labelId = `label-${id}`;
	return (
		<fieldset className="lunatic-component-set">
			<Legend id={labelId} description={description}>
				{legendText}
			</Legend>
			{children}
			<Errors errors={errors} activeId={id} />
		</fieldset>
	);
}

export default createCustomizableLunaticField(ComponentSet, 'ComponentSet');
