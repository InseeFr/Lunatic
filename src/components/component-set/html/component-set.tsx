import React, { PropsWithChildren, ReactNode } from 'react';
import { createCustomizableLunaticField, Errors } from '../../commons';
import Legend from './legend';
import { LunaticError } from '../../../use-lunatic/type';

type Props = PropsWithChildren<{
	id?: string;
	errors?: Record<string, LunaticError[]>;
	legendText: ReactNode;
	description?: ReactNode;
}>;

function ComponentSet({ id, errors, legendText, children }: Props) {
	const labelId = `label-${id}`;
	return (
		<fieldset className="lunatic-component-set">
			<Legend id={labelId}>{legendText}</Legend>
			{children}
			<Errors errors={errors} activeId={id} />
		</fieldset>
	);
}

export default createCustomizableLunaticField(ComponentSet, 'ComponentSet');
