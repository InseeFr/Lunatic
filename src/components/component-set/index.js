export { default } from './lunatic-component-set';
export { ComponentSetComponents } from './html/component-set-components';