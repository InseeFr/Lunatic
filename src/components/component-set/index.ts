export { default } from './lunatic-component-set';
