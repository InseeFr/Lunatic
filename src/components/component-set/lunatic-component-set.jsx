import React from 'react';
import LunaticComponent from '../commons/components/lunatic-component-without-label';
import ComponentSet from './html/component-set';
import ComponentSetComponents from './html/component-set-components';

function LunaticComponentSet(props) {
	const {
		declarations,
		label,
		id,
		components,
		value,
		executeExpression,
		shortcut,
		management,
		missing,
		features,
		preferences,
		errors,
		description,
		handleChange,
	} = props;

	return (
		<LunaticComponent
			id={id}
			preferences={preferences}
			declarations={declarations}
			value={value}
			missing={missing}
			management={management}
			description={description}
			components={components}
			executeExpression={executeExpression}
			shortcut={shortcut}
			features={features}
			handleChange={handleChange}
		>
			<ComponentSet
				id={id}
				legendText={label}
				errors={errors}
				components={components}
				value={value}
				features={features}
				executeExpression={executeExpression}
				preferences={preferences}
				management={management}
				missing={missing}
				shortcut={shortcut}
			>
				<ComponentSetComponents
					components={components}
					features={features}
					errors={errors}
					value={value}
					executeExpression={executeExpression}
					preferences={preferences}
					management={management}
					missing={missing}
					shortcut={shortcut}
					handleChange={handleChange}
				/>
			</ComponentSet>
		</LunaticComponent>
	);
}

export default LunaticComponentSet;
