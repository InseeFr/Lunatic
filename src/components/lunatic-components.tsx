import {
	Fragment,
	useRef,
	type PropsWithChildren,
	type ReactElement,
	type ReactNode,
} from 'react';
import { useAutoFocus } from '../hooks/use-auto-focus';
import { hasComponentType, hasLabel } from '../use-lunatic/commons/component';
import type { FilledLunaticComponentProps } from '../use-lunatic/commons/fill-components/fill-components';
import * as lunaticComponents from './index';

type Props<T extends Record<string, unknown>> = {
	// List of components to display (coming from getComponents)
	components: (
		| (FilledLunaticComponentProps & { colspan?: number; rowspan?: number })
		| { label: ReactElement; colspan?: number; rowspan?: number }
	)[];
	// Key that trigger autofocus when it changes (pageTag)
	autoFocusKey?: string;
	// Returns the list of extra props to add to components
	componentProps?: (
		component: FilledLunaticComponentProps & {
			colspan?: number;
			rowspan?: number;
		}
	) => T;
	// Add additional wrapper around each component
	wrapper?: (
		props: PropsWithChildren<
			FilledLunaticComponentProps &
				T & { index: number } & { colspan?: number; rowspan?: number }
		>
	) => ReactNode;
};

/**
 * Entry point for orchestrators, this component display the list of fields
 */
export function LunaticComponents<T extends Record<string, unknown>>({
	components,
	autoFocusKey,
	componentProps,
	wrapper = ({ children }) => (
		<div className="lunatic lunatic-component">{children}</div>
	),
}: Props<T>) {
	const wrapperRef = useRef<HTMLDivElement>(null);
	const hasComponents = components.length > 0;
	const WrapperComponent = autoFocusKey ? 'div' : Fragment;

	useAutoFocus(wrapperRef, hasComponents ? autoFocusKey : undefined);

	return (
		<WrapperComponent
			ref={WrapperComponent === Fragment ? undefined : wrapperRef}
		>
			{components.map((component, k) => {
				if (hasComponentType(component)) {
					const props = {
						...component,
						...componentProps?.(component),
					};
					return (
						<Fragment key={'id' in component ? component.id : `index-${k}`}>
							{wrapper({
								children: <LunaticComponent {...props} />,
								index: k,
								...props,
							})}
						</Fragment>
					);
				}

				if (hasLabel(component)) {
					const { label, ...props } = component;
					return (
						<Fragment key={`index-${k}`}>
							{wrapper({
								children: label,
								index: k,
								...props,
							} as any)}
						</Fragment>
					);
				}
				// In some case (table for instance) we have static component that only have a label (no componentType)
				return (
					<Fragment key={`index-${k}`}>
						{wrapper({
							children: component,
							index: k,
						} as any)}
					</Fragment>
				);
			})}
		</WrapperComponent>
	);
}

type ItemProps = FilledLunaticComponentProps;

function LunaticComponent(props: ItemProps) {
	// Component is too dynamic to be typed
	const Component = lunaticComponents[props.componentType] as any;
	return <Component {...props} />;
}
