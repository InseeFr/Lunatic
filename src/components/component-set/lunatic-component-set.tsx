import LunaticComponentWithoutLabel from '../commons/components/lunatic-component-without-label';
import type { LunaticComponentProps } from '../type';
import { LunaticComponents } from '../lunatic-components';
import { ComponentSet } from './html/component-set';
import { ComponentSetWrapper } from './html/component-set-wrapper';

export function LunaticComponentSet({
	id,
	label,
	declarations,
	description,
	response, // Do not propagate the response (it's undefined)
	components,
	...props // Props to propagate to children
}: LunaticComponentProps<'ComponentSet'>) {
	return (
		<LunaticComponentWithoutLabel
			id={id}
			declarations={declarations}
			description={description}
			preferences={props.preferences}
			value={props.value}
			missing={props.missing}
			management={props.management}
			handleChange={props.handleChange}
		>
			<ComponentSet id={id} legendText={label} errors={props.errors}>
				<LunaticComponents
					components={components}
					// Component props take precedence (we don't want to override value for instance)
					componentProps={(c) => ({ ...props, ...c })}
					wrapper={(p) => <ComponentSetWrapper {...p} />}
				/>
			</ComponentSet>
		</LunaticComponentWithoutLabel>
	);
}
