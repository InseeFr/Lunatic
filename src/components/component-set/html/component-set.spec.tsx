import { render } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { ComponentSetBase } from './component-set';
import { LunaticComponents } from '../../lunatic-components';
// import type { FilledLunaticComponentProps } from '../../../use-lunatic/commons/fill-components/fill-components';
import type { ReactNode } from 'react';

describe('ComponentSet', () => {
	it('renders without crashing', () => {
		const label = 'component set test';
		const components: any[] = [
			{
				label: 'label' as ReactNode,
				componentType: 'Subsequence',
				gotoPage: '1',
				conditionFilter: true,
				declarations: undefined,
				controls: undefined,
				shortcut: false,
				mandatory: false,
				preferences: ['COLLECTED'],
				id: 'toto',
			},
			{
				label: 'input label' as ReactNode,
				description: 'input description',
				componentType: 'Input',
				value: 'input value',
				conditionFilter: true,
				declarations: undefined,
				controls: undefined,
				shortcut: false,
				mandatory: false,
				preferences: ['COLLECTED'],
				id: 'toto1',
			},
		];
		const { container } = render(
			<ComponentSetBase
				legendText={label}
				id="componentSet"
				aria-labelledby="input"
			>
				<LunaticComponents components={components} />
			</ComponentSetBase>
		);
		expect(container).toMatchSnapshot();
	});

	it('should handle readOnly in sub-components', () => {
		const label = 'component set test';
		const components: any[] = [
			{
				label: 'label' as ReactNode,
				componentType: 'Subsequence',
				gotoPage: '1',
				conditionFilter: true,
				declarations: undefined,
				controls: undefined,
				shortcut: false,
				mandatory: false,
				preferences: ['COLLECTED'],
				id: 'toto',
			},
			{
				label: 'input label' as ReactNode,
				description: 'input description',
				componentType: 'Input',
				value: 'input value',
				conditionFilter: true,
				declarations: undefined,
				controls: undefined,
				shortcut: false,
				mandatory: false,
				preferences: ['COLLECTED'],
				id: 'toto1',
				readOnly: true,
			},
		];
		const { container } = render(
			<ComponentSetBase
				legendText={label}
				id="componentSet"
				aria-labelledby="input"
			>
				<LunaticComponents components={components} />
			</ComponentSetBase>
		);
		expect(container).toMatchSnapshot();

		const input = container.querySelector('input');
		expect(input).toHaveAttribute('readonly');
		(input as HTMLElement).focus();
		expect(input).toHaveFocus();
		expect(input).toHaveValue('input value');
	});
});
