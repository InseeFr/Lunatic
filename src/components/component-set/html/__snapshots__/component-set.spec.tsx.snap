// Vitest Snapshot v1

exports[`ComponentSet > renders without crashing 1`] = `
<div>
  <fieldset
    class="lunatic-component-set"
  >
    <legend
      class="lunatic-legend"
      id="label-componentSet"
    >
      component set test
    </legend>
    <div
      class="lunatic lunatic-component"
    >
      <div
        aria-label="subsequence-toto"
        class="subsequence-lunatic"
        id="subsequence-toto"
      >
        label
      </div>
    </div>
    <div
      class="lunatic lunatic-component"
    >
      <div
        class="field-container"
      >
        <div
          class="field"
        >
          <div
            class="lunatic-input"
          >
            <label
              class="lunatic-label"
              for="toto1"
              id="label-toto1"
            >
              input label
              <span
                class="label-description"
              >
                input description
              </span>
            </label>
            <input
              aria-labelledby="label-toto1"
              autocomplete="off"
              id="toto1"
              type="text"
              value="input value"
            />
          </div>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;

exports[`ComponentSet > should handle readOnly in sub-components 1`] = `
<div>
  <fieldset
    class="lunatic-component-set"
  >
    <legend
      class="lunatic-legend"
      id="label-componentSet"
    >
      component set test
    </legend>
    <div
      class="lunatic lunatic-component"
    >
      <div
        aria-label="subsequence-toto"
        class="subsequence-lunatic"
        id="subsequence-toto"
      >
        label
      </div>
    </div>
    <div
      class="lunatic lunatic-component"
    >
      <div
        class="field-container"
      >
        <div
          class="field"
        >
          <div
            class="lunatic-input"
          >
            <label
              class="lunatic-label"
              for="toto1"
              id="label-toto1"
            >
              input label
              <span
                class="label-description"
              >
                input description
              </span>
            </label>
            <input
              aria-labelledby="label-toto1"
              autocomplete="off"
              id="toto1"
              readonly=""
              type="text"
              value="input value"
            />
          </div>
        </div>
      </div>
    </div>
  </fieldset>
</div>
`;
